# Importing essential libraries

In [1]:
import selenium
import time
from selenium import webdriver

from bs4 import BeautifulSoup
import pandas as pd

# Loading the `Webdriver`

In [2]:
driver = webdriver.Chrome()

# Getting the access of `ACT Builder's` official website

In [2]:
driver.get('https://www.accesscanberra.act.gov.au/s/public-registers/construction-licences?registerid=m0I4a00000000CHEAY')
time.sleep(2)

NameError: name 'driver' is not defined

# Finding number of pages of the data 

In [112]:
pages=driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[6]/p').text
pages

'Page: 1 / 460'

In [113]:
numbers = []
for word in pages.split():
    if word.isdigit():
        numbers.append(int(word))
print(numbers[1])

460


# Main program to get `Headers` & `all the data rows` of just the `first page`

In [114]:
page_source = driver.page_source
soup =  BeautifulSoup(page_source, 'lxml')
            
headers = [i.text for i in soup.findAll('th')]
headers = headers[:-1]
print(headers)




data = [j.text for j in soup.findAll('td')]
    
all_names= data[::6]
    
data=data[1:]
all_license_no= data[::6]
    
data=data[1:]
all_exp= data[::6]
    
data=data[1:]
all_occupation= data[::6]
    
data=data[1:]
all_class= data[::6]

['Name', 'Licence No', 'Expiry Date', 'Occupation', 'Class Description']


# Main program to get `all the data rows` of all the pages `till 460`

In [115]:
for i in range(1, numbers[1]):
    
    driver.find_element_by_xpath('/html/body/div[4]/div[3]/div/div/div/div/div/div/div[3]/div/div/div[2]/c-cxs-public-registers-construction/div/div[6]/div[4]/button').click()
    
    page_source = driver.page_source
    soup =  BeautifulSoup(page_source, 'lxml')
    
    data = [j.text for j in soup.findAll('td')]
    
    name= data[::6]
    #name
    all_names= all_names+ name
    
    data=data[1:]
    lic_no= data[::6]
    #lic_no
    all_license_no=all_license_no+ lic_no
    
    data=data[1:]
    exp= data[::6]
    #exp
    all_exp=all_exp+exp
    
    data=data[1:]
    occupation= data[::6]
    #occupation
    all_occupation=all_occupation+occupation
    
    data=data[1:]
    class_build= data[::6]
    #class_build
    all_class=all_class+class_build

# Checking number of records we retrieved

In [116]:
len(all_names)

4596

## Name of the first record in our `ACT Builder dataset`

In [117]:
all_names[0]

'2NV CONSTRUCTIONS PTY LTD'

## Name of the last record in our `ACT Builder dataset`

In [118]:
all_names[-1]

'ZYCON BUILDING PTY LTD'

# Creating the final dataframe by zipping all the attribute's lists

In [119]:
df = pd.DataFrame(list(zip(all_names, all_license_no, all_exp, all_occupation, all_class)),
               columns =['Name', 'License Number','Expiry date','Occupation','Class'])
df

Name License Number  \
0                    2NV CONSTRUCTIONS PTY LTD        2019305   
1                           35 DEGREES PTY LTD       20151084   
2                                   37 PTY LTD       20201039   
3                A & A CONTRACTORS PTY LIMITED       19968065   
4                    A & B DAL CORTIVO PTY LTD        2008265   
5                          A & D CONSTRUCTIONS       19894869   
6           A & E PROPERTY MAINTENANCE PTY LTD       20181082   
7                          A & J HOMES PTY LTD        2020562   
8         A & J INVESTMENTS (CANBERRA) PTY LTD       20121515   
9             A & J PROJECTS AUSTRALIA PTY LTD        2017986   
10                       A & N & D APOSTOLOSKI        2017920   
11      A & P PARKES CONSTRUCTIONS PTY LIMITED        2009381   
12                 A & T CONSTRUCTIONS PTY LTD      200213079   
13                   A FILIPOSKI & S FILIPOSKI        2020484   
14               A PLUS PLUMBING (ACT) PTY LTD        2016683   
15                          A SARRI & DA SARRI        2019451   
16                       A-SHELL HOMES PTY LTD        2010325   
17                  A.C.N. 109 626 280 PTY LTD        2015306   
18                           A.J. EWER PTY LTD      200413692   
19                 A.J. GRANT BUILDING PTY LTD        2007200   
20           A1 ADVANCED CONSTRUCTIONS PTY LTD        2021836   
21                   A1 BUILDING GROUP PTY LTD        2015611   
22                       A2Z HOMES ACT PTY LTD        2021900   
23         AA BUILDING AND MAINTENANCE PTY LTD         201685   
24                      AABRE BUILDING PTY LTD        2020204   
25    ABA CONSTRUCTION MANAGERS (AUST) PTY LTD       20121196   
26                              ABALTA PTY LTD      199811529   
27                        ABBOTT, STUART DAVID         201260   
28                                ABBOTT, THEO        2020807   
29                     ABC CONSTRUCTIONS - ACT      200427033   
...                                        ...            ...   
4566                 ZAUNER, GARRY CHRISTOPHER        2013379   
4567                         ZCN HOMES PTY LTD       20191180   
4568                        ZDRAVEVSKI, STEFAN        2019368   
4569                             ZEC, ANGELINA       20191185   
4570                              ZEC, MILOMIR       20121156   
4571                               ZEC, SLAVKO       20121155   
4572                        ZEINEDDINE, SALMAN         201786   
4573                         ZEITHHOFER, CHRIS       20181236   
4574                            ZELLER, HELMUT       19802058   
4575                           ZERELLA, COSIMO        2021923   
4576                        ZETA HOMES PTY LTD       20191156   
4577                               ZHANG, BIAO        2021518   
4578                               ZHANG, JING         201232   
4579                                 ZHANG, QI        2014343   
4580                                 ZHANG, QI        2014442   
4581                              ZHANG, RUIBO        2020587   
4582                                ZHAO, MENG        2021483   
4583                              ZHAO, YOUXIN        2017279   
4584                               ZHENG, JIEN        2020767   
4585                               ZHENG, QIAN        2018647   
4586                              ZHOU, ZEFENG        2019740   
4587                         ZHU, XUHUA GEORGE       20211264   
4588                               ZIA-UD, DIN        2021409   
4589                     ZIEGLER, MICHAEL JOHN      199610673   
4590                              ZILIC, MARKO       20111246   
4591                               ZIVKO, IVAN       19884304   
4592                        ZMISA, SASHA DAVID        2019547   
4593                              ZOVAK, DRAGO        2019232   
4594                    ZUBOVIC, ZELJKO THOMAS       19946847   
4595                    ZYCON BUILDING PTY LTD        2016957   

           

## Checking the shape (rows & columns) of the final dataframe

In [120]:
df.shape

(4596, 5)

# Converting the dataframe into a `.csv` file

In [121]:
df.to_csv('ACT_Builders.csv')